arb_os/output.mini

```diff
@@ -343,14 +344,61 @@ public impure func queueMessageForSend(msg: ByteArray) {
     evmCallStack_queueMessage(msg);
 }
 
-public impure func sendQueuedMessages(queue: Queue) {
-    loop {
+public impure func sendQueuedMessages(queue: Queue, evmLogs: EvmLogs) -> EvmLogs {
+    return xloop {
         if let Some(res) = queue_get(queue) {
-            let (uq, msg) = res;
+            let (uq, rawMsg) = res;
             queue = uq;
-            outbox_append(unsafecast<ByteArray>(msg));
+            let msg = unsafecast<ByteArray>(rawMsg);
+            if (bytearray_getByte(unsafecast<ByteArray>(msg), 0) == const::SendType_sendTxToL1) {
+                // we need to find the EVM log corresponding to this send, and fill in batch number, item num in batch
+                evmLogs = evmlogs_forall(
+                    evmLogs,
+                    updateEvmLogForSendToL1,
+                    true,
+                ).0;
+            }
+            outbox_append(msg);
         } else {
-            return;
+            return evmLogs;
         }
-    }
+    };
+}
+
+type LogWith3Topics = struct {
+    sender: address,
+    data: struct {
+        size: uint,
+        buf: buffer
+    },
+    topic0: uint,
+    topic1: uint,
+    topic2: uint,
+}
+
+impure func updateEvmLogForSendToL1(rawLog: any, state: any) -> (any, any) {
+    // This is designed to be used by evmlogs_forall. When used by evmlogs_forall in its pass through
+    //      the current set of logs, this morphs the first EVM log with topic[0] == const::EVMLogTopicPlaceHolderForL2ToL1Send
+    //      into a proper EVM log item with batch-related fields filled in
+
+    return xif(unsafecast<bool>(state)) {
+        // Cast the log item to a struct that assumes 3 topics. topic1 and topic2 might not exist,
+        //      but that's OK because we won't try to access them until after we have verified that they exist.
+        let log = unsafecast<LogWith3Topics>(rawLog);
+        xif (log.topic0 == const::EVMLogTopicPlaceHolderForL2ToL1Send) {
```

Comment:
I think 0 topic logs will break this. Can we check the tuple length first to make sure it's big enough?

Reply:
Added a length check.